---
title: "Simulating microbiome data based on SparseDOSSA2 package"
author: 
- name: Fentaw Abegaz
  affiliation: BDA, University of Amsterdam, Netherlands
  email: f.a.yazew@uva.nl
date: 'Compiled: `r format(Sys.Date(), "%d %B, %Y")`'
output: 
  html_document:  
    toc: yes
    #toc_float: yes
    number_section: yes
    highlight: tango
    theme: united
pdf_document:
    toc: yes
vignette: >
  %\VignetteEcoding{UTF-8}
  %\VignetteIndexEntry{Vignette Title}
  %\VignetteIndexEntry{Intro}
  #\usepackage[utf8]{inputenc}
  %\VignetteEngine{knitr::rmarkdown}
---

```{r options, include=FALSE, echo=FALSE}
knitr::opts_chunk$set(
    warning = FALSE, 
    error = FALSE, 
    message = FALSE)
```
# Load all the required libraries and functions

## Set working directory

```{r setup}

workingDirectory <- getwd()
```


## Load all the required libraries
```{r, echo=FALSE}
library(pacman)
library(devtools)
library(pkgmaker)
library(optparse)
library(parallel)
library(stringi)
library(doParallel)
library(SparseDOSSA2)
library(zinbwave)
library(SummarizedExperiment)
library(dplyr)
library(readr)
library(tibble)
library(BiocParallel)
library(edgeR)
library(DESeq2)
library(phyloseq)
library(ggplot2)
library(ape)
library(miceadds)
```

## Load simulation functions Functions
The functions are included in the folder "\\sim_func".

```{r, echo=FALSE}
Rsfolder_pathSim <- ".\\sim_func"
source.all(Rsfolder_pathSim, grepstring="\\.R",  print.source=TRUE )

```
## Load template data in phyloseq format
Here we used a subset of the metagenome shotgun sequencing data from the Human Microbiome Project (HMP-2012), which include 5 supragingival and 5 subgingival plaque samples from the oral cavity[1, 2].

1. 	Calgaro M, Romualdi C, Waldron L, Risso D, Vitulo N. Assessment of statistical methods from single cell,       bulk RNA-seq, and metagenomics applied to microbiome data. Genome Biol. 2020;21: 191.                  
    doi:10.1186/s13059-020-02104-1

2. 	The Human Microbiome Project Consortium. Structure, function and diversity of the healthy human microbiome.     Nature. 2012;486: 207-214. doi:10.1038/nature11234

```{r, echo=FALSE}
physeqLOAD <- readRDS("..\\data\\subANDsup_gingival_species.RDS")

grp1 = "subgingival_plaque"
grp2 = "supragingival_plaque"
physeqLOAD@sam_data$group = factor(physeqLOAD@sam_data$body_subsite,levels = c(grp1,grp2),labels = c("grp1","grp2"))
taxadata <- tax_table(physeqLOAD)  

counts <- t(otu_table(physeqLOAD, 'matrix'))
group <- sample_data(physeqLOAD)$group
#taxadata <- tax_table(physeqLOAD)  

# #Template microbiome datasets used in the paper
# #1. Forest_potting soils experimental data 
# physeqLOAD <- readRDS("..\\data\\soil_sample_gr2.rds")
# sample_data(physeqLOAD)$group <- sample_data(physeqLOAD)$Soil_Type
# 
# #2. N-P starvation experimental data 
# physeqLOAD <- readRDS(file = "..\\data\\Davar_cleaned_DataHoaglandCompartment.RDS")
# sample_data(physeqLOAD)$group <- sample_data(physeqLOAD)$sample_title

```

## Clean the overall data appropreiate for SparseDOSSA2 simulation

```{r, echo=FALSE}
cleaned_taxa <- clean_data_sim(physeqLOAD)
filtered_real_counts <- cleaned_taxa$filtered_real_counts

readDepth <-  cleaned_taxa$readDepth
cat("Read depth=", readDepth , "\n") 
nMicrobes  <- cleaned_taxa$nMicrobes
cat("number of microbes=", nMicrobes,  "\n")

features_otu <- cleaned_taxa$features_otu
cat("filtered number of taxa without groupwise structured zeros=", ncol(features_otu), "\n") 
zeroGroup_taxa <- cleaned_taxa$zeroGroup_features
if(is.null(zeroGroup_taxa)) cat("number of taxa with groupwise structured zeros excluded from the simulation studies=", 0,  "\n") else cat("number of taxa with groupwise structured zeros excluded from comparing differential abundance methods =", ncol(zeroGroup_taxa),  "\n") 
```
# Simulations

## Set parameters for the simulations
```{r, echo=FALSE}
#Set the number of spike-in and randomly sample from the spike-in candidates
nspike<- 30 #70 #50 20 #(lessthan or equal to ncol(features_otu))

#set a vector of log-fold changes
es <- c(0.5,1, 2)
#set a vector of total sample sizes
ns <- c(10,20,50,100,200)

#Number of simulations
nIterations <- 100
```


## Run simulation using SparseDOSSA2 (https://huttenhower.sph.harvard.edu/sparsedossa2) on the template data. The simulated dataset is stored in a folder "Input".
```{r, echo=FALSE}
input_simulated_data <- run_simulator_SparseDOSSA2(es, ns, nIterations, workingDirectory,
                                                   nMicrobes,readDepth,filtered_real_counts)

```
